handler/datasource/pushitem.go

```diff
@@ -84,3 +88,157 @@ func pushItemHandler(
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to check existing item")
+	}
+	if existing > 0 {
+		logger.Debugw("item already exists", "source_id", source.ID, "path", obj.Remote())
+		return nil, nil, nil
+	}
+	item := model.Item{
+		SourceID:                  source.ID,
+		Path:                      obj.Remote(),
+		Size:                      size,
+		LastModifiedTimestampNano: lastModified.UnixNano(),
+		Hash:                      hashValue,
+	}
+	logger.Debugw("new item", "item", item)
+	err = EnsureParentDirectories(db, &item, rootID, directoryCache)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to ensure parent directories")
+	}
+	var itemParts []model.ItemPart
+	err = database.DoRetry(func() error {
+		return db.Transaction(func(db *gorm.DB) error {
+			err := db.Create(&item).Error
+			if err != nil {
+				return errors.Wrap(err, "failed to create item")
+			}
+			if dataset.UseEncryption() {
+				itemParts = append(itemParts, model.ItemPart{
+					ItemID: item.ID,
+					Offset: 0,
+					Length: item.Size,
+				})
+			} else {
+				offset := int64(0)
+				for {
+					length := splitSize
+					if item.Size-offset < length {
+						length = item.Size - offset
+					}
+					itemParts = append(itemParts, model.ItemPart{
+						ItemID: item.ID,
+						Offset: offset,
+						Length: length,
+					})
+					offset += length
+					if offset >= item.Size {
+						break
+					}
+				}
+			}
+			err = db.Create(&itemParts).Error
+			if err != nil {
+				return errors.Wrap(err, "failed to create item parts")
+			}
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create item")
+	}
+	return &item, itemParts, nil
+}
+
+func EnsureParentDirectories(db *gorm.DB,
+	item *model.Item, rootDirID uint64,
+	directoryCache map[string]uint64) error {
+	if item.DirectoryID != nil {
+		return nil
+	}
+	last := rootDirID
+	segments := strings.Split(item.Path, "/")
+	for i, segment := range segments[:len(segments)-1] {
+		p := strings.Join(segments[:i+1], "/")
+		if dirID, ok := directoryCache[p]; ok {
+			last = dirID
+			continue
+		}
+		newDir := model.Directory{
+			SourceID: item.SourceID,
+			Name:     segment,
+			ParentID: &last,
+		}
+		logger.Debugw("creating directory", "path", p, "dir", newDir)
+		err := database.DoRetry(func() error {
+			return db.Transaction(func(db *gorm.DB) error {
+				return db.
+					Where("parent_id = ? AND name = ?", last, segment).
+					FirstOrCreate(&newDir).Error
+			})
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to create directory")
+		}
+		directoryCache[p] = newDir.ID
+		last = newDir.ID
+	}
+
+	item.DirectoryID = &last
+	return nil
+}
+
+func MaxSizeToSplitSize(m int64) int64 {
+	r := util.NextPowerOfTwo(uint64(m)) / 4
+	if r > 1<<30 {
+		r = 1 << 30
+	}
+
+	return int64(r)
+}
+
+func ExtractFromFsObject(ctx context.Context, info fs.ObjectInfo) (size int64, hashValue string, lastModified time.Time, lastModifiedReliable bool) {
```

Comment:
seems like this method should be able to return an error?

Reply:
i just moved it verbatim from another file, but i could add that functionality as a part of this pr if it would be useful